src/ripple/nodestore/ShardStoragePaths.md

```diff
@@ -0,0 +1,56 @@
+
+## Configuration
+
+The `shard_db` and `historical_shard_paths` sections of the config file will be used to determine where the server stores shards. Minimally, the `shard_db` section must contain a single `main_storage_path` key. If this is the only storage path provided, all shards will be stored at this location. If the configuration also lists one or more lines in the `historical_shard_paths` section, all older shards will be stored at these locations, and the `main_storage_path` will be used only to store the current and previous shards. The goal is to allow users to provide an efficient SSD for storing recent shards, as these will be accessed more frequently, while using large mechanical drives for storing older shards that will be accessed less frequently.
+
+Below is a sample configuration snippet that provides a path for main storage and several paths for historical storage:
+
+```dosini
+# This is the persistent datastore for shards. It is important for the health
+# of the ripple network that rippled operators shard as much as practical.
+# NuDB requires SSD storage. Helpful information can be found here
+# https://ripple.com/build/history-sharding
+[shard_db]
+type=NuDB
+
+# A single path for storing
+# the current and previous
+# shards:
+# -------------------------
+main_storage_path=/var/lib/rippled/db/shards/nudb
+
+# Path where shards are stored
+# while being downloaded:
+# ----------------------------
+partial_shard_download_path=/var/lib/rippled/db/shards/
+
+# The number of historical shards to
+# store. The default value is 0, which
+# means that only the current and
+# previous shards will be stored.
+# ------------------------------------
+max_historical_shards=100
+
+# List of paths for storing older shards.
+[historical_shard_paths]
+/dev/sdb1
+/dev/sdb2
+/dev/sdb3
+
+```
+## Shard Migration
+
+When a new shard (*current shard*) is confirmed by the network, the recent shards will shift. The next shard to build will become the *current shard* and the *current shard* will become the *previous shard*. These are just logical labels, and the shards themselves don't change to reflect being current or previous. However, if the server's configuration specifies one or more paths for historical storage, during this shift the formerly *previous shard* will be migrated to one of the historical paths. If multiple paths are provided, the server dynamically chooses one with sufficient space for storing the shard.
+
+### Selecting a Historical Storage Path
+
+When storing historical shards, if multiple historical paths are provided, the paths will be selected in a round-robin fashion. By using all available storage devices, we increase our chances for concurrent access to stored shards. Naturally, a storage path is included in the round-robin distribution only if it has enough storage capacity to hold the next shard.
```

Comment:
This would cause issues if I for example have 3 paths configured, the respective disks start getting full and I add a 4th one. I would recommend to store the next shard at the disk that has the least space used by shards (still an issue if disk sizes are different, but less problematic than the "proper" solution of choosing the disk that has the most free space) - or, if getting sizes is too difficult, then at least choosing the disk that has the least amount of shard files stored, with a random selection (or just the order in the config file) as tie-breaker.

Reply:
The selection process outlined here is actually outdated - during the implementation I decided to go with random selection. The doc in the most recent push has been updated with that change.

I think the random selection is sufficient for our needs. It's simple (which is good) and does the job well enough. If the motivation for using capacity-based selection is to increase performance by reducing file fragmentation, I don't think its worth implementing. For any server that stores limited history, shard fragmentation is inevitable and eventually all of its historical shards will be fragmented. And for a server that stores full history, in theory shard fragmentation won't occur. In either case I don't see a reason for space-based selection.